Adds support for sslyze>=3.0.0

SaptakS opened the pull request:
Updates the https_check function to use code that works with sslyze>=3.0.0

- fixes import errors for the current required modules
- uses the new format for certificate analyzer
- updates dependency of sslyze to >=3.0.0

Refs: #209 

Also, thanks to [Ethan](https://github.com/Ethanljf)'s [code snippet](https://github.com/cisagov/pshtt/issues/209#issue-591146621) that helped in this commit.

## 🧪 Testing ##

Tested the code using the current test suite of pshtt in debian 10, with python 3.7
Following is the command I ran: `python3 -m pytest`

## ✅ Checklist ##

* [x] This PR has an informative and human-readable title.
* [x] Changes are limited to a single goal - _eschew scope creep!_
* [x] All relevant type-of-change labels have been added.
* [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
* [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
* [x] All new and existing tests pass.

SaptakS committed:
Adds support for sslyze>=3.0.0

- fixes import errors for the current required modules
- uses the new format for certificate analyzer
- updates dependency of sslyze to >=3.0.0

SaptakS committed:
Adds extra certificate informations for user trusted CA

hillaryj commented on pshtt/pshtt.py:
```diff
@@ -654,88 +685,51 @@ def https_check(endpoint):
+    # Check certificate chain
+    # NOTE: If this is the only flag that's set, it's probably
+    # an incomplete chain
+    # If this isnt the only flag that is set, it's might be
```
One small fix in a (previously-existing) comment as a suggested change:

```suggestion
    # If this isn't the only flag that is set, it might be
```

felddy replied:
Fixed in 81882c488e099cd96e1525e62e275cf7759f506f

hillaryj reviewed (approved):
Thanks for the contribution!

I've tested via the python tests as well as ran the tool locally (Mac OSX Catalina with Python 3.8.3) and got a nice little results.csv file out. Several other folks are reviewing as well - great work!

The codebase needs to be `black`ened as well as [skeletonized](https://github.com/cisagov/development-guide/tree/develop/project_setup) and I'll add an issue for that if one doesn't exist.

Thanks again and great work!

jsf9k reviewed (approved):
This all looks great to me.  A big thank you to @SaptakS and @Ethanljf for working on this!

jsf9k commented:
I think we should do a minor version bump before merging this.  I tried to do this myself, but I can't commit to this branch, so I think @SaptakS will have to do it.  Here's how:
1. Make sure you have the `semver` Python package installed
2. From the `pshtt` project root, simply run `./bump_version.sh minor`

mcdonnnj commented:
> 
> 
> I think we should do a minor version bump before merging this. I tried to do this myself, but I can't commit to this branch, so I think @SaptakS will have to do it. Here's how:
> 
>     1. Make sure you have the `semver` Python package installed
> 
>     2. From the `pshtt` project root, simply run `./bump_version.sh minor`

GitHub has a feature to allow maintainers to edit forks:
https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

SaptakS committed:
Fixes typo in comment

SaptakS commented:
I agree with @mcdonnnj that the maintainers with write access should have permission to push to branches of fork. But I have done the version bump and the typo fix in the comment.

cc: @jsf9k @hillaryj

SaptakS commented:
> The codebase needs to be blackened as well as skeletonized and I'll add an issue for that if one doesn't exist.

I haven't done this in the PR because `black` is making changes in a lot of unrelated files.

mcdonnnj commented:
> 
> 
> > The codebase needs to be blackened as well as skeletonized and I'll add an issue for that if one doesn't exist.
> 
> I haven't done this in the PR because `black` is making changes in a lot of unrelated files.

That's fine as that is an "us" task, so don't worry about that. It's a long-term maintenance task for this project that we just haven't had the time to do yet.

hillaryj commented:
Yup, that's absolutely out of the scope of this PR, although thank you for trying!

I added follow-on task https://github.com/cisagov/pshtt/issues/215 Skeletonize repository and standardize code formatting.

SaptakS commented:
Hi @mcdonnnj , is there anything else needed from my side on this PR?

jsf9k commented:
> Hi @mcdonnnj , is there anything else needed from my side on this PR?

Nope, I think we are just waiting on reviews.

@mcdonnnj, are you good with merging once we get a positive review from you or is there someone else you feel we should wait for?

mcdonnnj commented:
> 
> 
> > Hi @mcdonnnj , is there anything else needed from my side on this PR?
> 
> Nope, I think we are just waiting on reviews.
> 
> @mcdonnnj, are you good with merging once we get a positive review from you or is there someone else you feel we should wait for?

Review is on my to-do, I will finish it tonight or tomorrow morning. @jsf9k did we want to get anything from @h-m-f-t? Other than that there are no additional blockers from me.

h-m-f-t commented:
I'm good here and don't need to review. Thanks @SaptakS!

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -594,14 +605,34 @@ def https_check(endpoint):
+        command = ScanCommand.CERTIFICATE_INFO
+        if CA_FILE is not None:
+            command_extra_args = {
+                ScanCommand.CERTIFICATE_INFO: CertificateInfoExtraArguments(custom_ca_file=Path(CA_FILE))
```
`command` is somewhat superfluous now that it is not something we have to create, but if we're using it, then it makes sense to tie it in here at least.

```suggestion
                command: CertificateInfoExtraArguments(custom_ca_file=Path(CA_FILE))
```

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -623,7 +654,7 @@ def https_check(endpoint):
         custom_trust = True
         public_not_trusted_string = ""
-        validation_results = cert_plugin_result.path_validation_result_list
+        validation_results = cert_plugin_result.certificate_deployments[0].path_validation_results
```
@jsf9k @h-m-f-t Should we pull in results for *all* detected leaf certificates, or just the first, and typically only, certificate in this list? I'm not sure how often this would impact our scoring, but thought it worth mentioning.

More information about `certificate_deployments` is in the [attribute comments in `sslyze`](https://github.com/nabla-c0d3/sslyze/blob/e45cb661bf983bd44f6172a3f48a1386155c9814/sslyze/plugins/certificate_info/implementation.py#L43-L45).

h-m-f-t replied:
As a user, I'd prefer a tool be inclusive rather than presume there's only ever one cert – but without doing a run of our namespace I'm similarly unsure whether this would impact folks we scan.

SaptakS replied:
@mcdonnnj I can just loop over all the results just to be safe. Given most times there will be only one certificate, the loop will get executed only once, resulting in a similar code flow like this one. What do you think?

SaptakS replied:
@mcdonnnj in case of multiple leaf certificates, do we consider trusted if at least one has validations_results successful? Or do we consider trust if all leaf certificates have validation_results successful?

mcdonnnj replied:
@SaptakS I'm waiting for a team consensus on which way we want to lean on multiple certificates. I would hesitantly say that if we're checking all leaf certificates, then they all must be successful to be trusted, but that's my stance.

SaptakS replied:
I have currently implemented it using that logic in the latest commit. So do check that out, if that works.

jsf9k replied:
I agree with what @mcdonnnj said - all leaf certificates must be successful to be trusted, which is what you have implemented.

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -654,88 +685,51 @@ def https_check(endpoint):
+        endpoint.https_expired_cert = True
+
+    # Check to see if the cert is self-signed
+    if leaf_cert.issuer is leaf_cert.subject:
```
The `is` comparison will only work if `cert.issuer` and `cert.subject` are the same object. Using `self-signed.badssl.com` to test the values of a cert:

```console
>>> ss_leaf.issuer
<Name(C=US,ST=California,L=San Francisco,O=BadSSL,CN=*.badssl.com)>
>>> ss_leaf.subject
<Name(C=US,ST=California,L=San Francisco,O=BadSSL,CN=*.badssl.com)>
>>> ss_leaf.issuer is ss_leaf.subject
False
>>> ss_leaf.issuer == ss_leaf.subject
True
```

```suggestion
    if leaf_cert.issuer == leaf_cert.subject:
```

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -654,88 +685,51 @@ def https_check(endpoint):
-                (("self signed certificate") in msg)
-            )
-        ):
+        if cert.issuer is (cert.subject or None):
```
Same comment as above about usage of `is`. I'm not a big fan of a `cert.issuer == None` possibility, so it could be refactored.

```suggestion
        if cert.issuer == (cert.subject or None):
```

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -748,9 +742,21 @@ def https_check(endpoint):
+                        scanner = Scanner()
+                        command = ScanCommand.CERTIFICATE_INFO
+                        command_extra_args = {
+                            ScanCommand.CERTIFICATE_INFO: CertificateInfoExtraArguments(custom_ca_file=Path(PT_INT_CA_FILE))
```
Same comments as previously.

```suggestion
                            command: CertificateInfoExtraArguments(custom_ca_file=Path(PT_INT_CA_FILE))
```

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -654,88 +685,51 @@ def https_check(endpoint):
         if (
                 endpoint.https_self_signed_cert is False and (
-                    len(cert_plugin_result.received_certificate_chain) < 2
+                    len(cert_plugin_result.certificate_deployments[0].received_certificate_chain) < 2
```
```suggestion
                    endpoint.https_cert_chain_len < 2
```

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -594,14 +605,34 @@ def https_check(endpoint):
                 logging.exception("{}: Retrying sslyze scanner certificate plugin.".format(endpoint.url))
-                cert_plugin_result = scanner.run_scan_command(server_info, command)
+                scanner.queue_scan(scan_request)
+                # Retrieve results from generator object
```
Just wanted to highlight that we're consuming the iterator, and we are only using the first result. This should be fine for how this library does scanning, but I still think it's worth explicitly commenting.

```suggestion
                # Consume the generator object and retrieve the first result
```

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -748,9 +742,21 @@ def https_check(endpoint):
+                            scan_commands=[command]
+                        )
+                        scanner.queue_scan(scan_request)
+                        # Retrieve results from generator object
```
Same rationale as the earlier feedback.

```suggestion
                        # Consume the generator object and retrieve the first result
```

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -654,88 +685,51 @@ def https_check(endpoint):
-            (("certificate has expired") in msg)
-        ):
-            endpoint.https_expired_cert = True
+    cert_chain = cert_plugin_result.certificate_deployments[0].received_certificate_chain
```
Using `.received_certificate_chain` instead of using `.path_validation_results`, filtering on `.trust_store.name == STORE` for each element of that list, and then using `.verified_certificate_chain` would ignore the usage of `STORE` before this PR in these checks that follow. We care about the result from the `Mozilla` trust store, or a `Custom` store that is provided.

@jsf9k I would really like your thoughts here.

SaptakS replied:
I think going `.path_validation_results` sounds like it will be more consistent with functionalities before this PR. So maybe I will just implement that flow and check with `.trust_store.name` to take the STORE into consideration. cc @mcdonnnj @jsf9k

SaptakS replied:
@mcdonnnj like in the other comment, should we consider all certificate deployments here as well?

jsf9k replied:
@SaptakS - yes, let's keep it consistent with the functionality before the PR.  Let's also consider all certificate deployments.

mcdonnnj reviewed (changes requested):
I think I hit all the points I saw. I would like feedback to some of the questions I asked below about the impact of some of these changes and on if we care about *every* leaf certificate detected, or just the first one in the list.

SaptakS committed:
Add fixes and suggestions from code review

- fix comments
- use == instead of is

Co-authored-by: Nick M. <50747025+mcdonnnj@users.noreply.github.com>

SaptakS committed:
Consider all certificate deployments instead of first one

- consider all deployments and if all validations succeed, then
consider it trusted
- use path_validation_results instead of received_certificate_chain
to consider if STORE is included in trust_store
- check self signed validation in cert_chain till the second last
cert since the last cert is root cert and hence is self signed anyways
- put all logic inside one loop over path_validation_results

SaptakS commented:
@mcdonnnj @jsf9k @h-m-f-t I made some updates in the code such that it takes into consideration all the certificate_deployments instead of just the first, and also checks if `STORE` is in `trust_store.name` of `path_validation_results` before assessing https_validity check.

Also, I realized there was a bug in my code, where it was checking the root certificate in the chain for self signed, returning true, and then declaring the https chain to be bad. So I fixed that so that we check till the second last certificate in the cert_chain.

@mcdonnnj let me know if this fixes all the requested changes.

SaptakS commented:
Hi @mcdonnnj. Any updates on this?

jsf9k commented:
> Hi @mcdonnnj. Any updates on this?

@SaptakS, let me first apologize for taking so long to re-review this PR.  We deeply value your contribution, but I am being pulled in many different directions right now and my time management skills need some work.  If you can make the change to consider all certificate deployments (near line 700 of `pshtt.py`) then I think this PR will be good to go.

I also dig your avatar.  It makes me think of Jerry Garcia and Allen Ginsberg.  :peace_symbol:

SaptakS commented:
> If you can make the change to consider all certificate deployments (near line 700 of pshtt.py) then I think this PR will be good to go.

I am not sure I fully understand 😅 . I made the changes to consider all certificate deployments. If you are referring to [this line](https://github.com/cisagov/pshtt/pull/214/files#diff-4cb7eb100f80cc3b3589ab91fe6f4b1cfef83929fac6dc075f23696a84c75711R700) I am not checking if the last cert is self-signed or not, since it's a root certificate.

Do you mean in that case I should check for certificate expiration for all certs, but avoid the self signing check on the last cert? As far as I know root certs will be self-signed, right? @mcdonnnj 

> I also dig your avatar. It makes me think of Jerry Garcia and Allen Ginsberg.

Thanks!

jsf9k commented:
> I am not sure I fully understand  . I made the changes to consider all certificate deployments. If you are referring to [this line](https://github.com/cisagov/pshtt/pull/214/files#diff-4cb7eb100f80cc3b3589ab91fe6f4b1cfef83929fac6dc075f23696a84c75711R700) I am not checking if the last cert is self-signed or not, since it's a root certificate.
> 
> Do you mean in that case I should check for certificate expiration for all certs, but avoid the self signing check on the last cert? As far as I know root certs will be self-signed, right? @mcdonnnj

I just meant to do the things you mentioned in your comments [here](https://github.com/cisagov/pshtt/pull/214#discussion_r569671237) and [here](https://github.com/cisagov/pshtt/pull/214#discussion_r569751196).  You're right - it looks like you already did those things.  I was paying more attention to the PR comments than to the code - mea culpa.

jsf9k reviewed (approved):
Thank you @SaptakS for making these changes, and especially for being patient with us during the review process!

eloquence commented:
Hi folks, over at @freedomofpress we're hoping to land this PR in order to keep the `cryptography` dependency up-to-date for repos such as https://github.com/freedomofpress/securethenews/ . It looks like it's been approved and all code review comments have been addressed - is there anything we can do to help move it along from here?

mcdonnnj commented:
> 
> 
> Hi folks, over at @freedomofpress we're hoping to land this PR in order to keep the `cryptography` dependency up-to-date for repos such as https://github.com/freedomofpress/securethenews/ . It looks like it's been approved and all code review comments have been addressed - is there anything we can do to help move it along from here?

Apologies for this falling off the radar a bit. I need to do one last sweep this week with the final determination on multiple certificate chains in mind to make sure everything lines up. I also wanted to do some last spot testing to ensure there aren't any immediate issues around how we use this package.

So final resolution should be end of the week at the latest.

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -568,19 +576,22 @@ def https_check(endpoint):
+    except ConnectionToServerFailed as err:
         endpoint.live = False
         endpoint.https_valid = False
         logging.exception("{}: Error in sslyze server connectivity check when connecting to {}".format(endpoint.url, err.server_info.hostname))
```
This has changed in `sslyze`:

```suggestion
        logging.exception("{}: Error in sslyze server connectivity check when connecting to {}".format(endpoint.url, err.server_location.hostname))
```

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -651,106 +728,41 @@ def https_check(endpoint):
 
     try:
-        endpoint.https_cert_chain_len = len(cert_plugin_result.received_certificate_chain)
+        endpoint.https_cert_chain_len = len(cert_plugin_result.certificate_deployments[0].received_certificate_chain)
```
@jsf9k Thoughts on how we should handle multiple certificate deployments? Running this against the [current federal domain list](https://github.com/GSA/data/blob/master/dotgov-domains/current-federal.csv) there were at least ~50 domains that have two deployments.

SaptakS replied:
@mcdonnnj I can loop over all the certificate_deployments and add the length of received_certificate_chain. I think that should address this?

mcdonnnj replied:
@SaptakS This is another one that I'd like consensus on our end. The problem I see with just adding lengths is that it does not give clear insight into what the value means anymore. Does a value of `4` mean a single chain with a length of four, or does it mean two chains of length two?

mcdonnnj replied:
I'll add a mention for @h-m-f-t to get more thoughts.

h-m-f-t replied:
For now, I'm not especially bothered by the value representation as it stands. Simply adding the length would likely provide enough detail should we need to inspect longer chains because the outliers will be few and will be presented by a higher number.

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -651,106 +728,41 @@ def https_check(endpoint):
+                        # Consume the generator object and retrieve the first result
+                        scan_result = [x for x in scanner.get_results()][0]
+                        cert_plugin_result = scan_result.scan_commands_results[ScanCommand.CERTIFICATE_INFO]
+                        if(cert_plugin_result.certificate_deployments[0].verified_certificate_chain is not None):
```
Another question of how we want to handle multiple certificate deployments here as well.

mcdonnnj commented on pshtt/pshtt.py:
```diff
@@ -651,106 +728,41 @@ def https_check(endpoint):
             # *** TODO check that it is not a bad hostname and that the root cert is trusted before suggesting that it is an intermediate cert issue.
             endpoint.https_missing_intermediate_cert = True
-            if(cert_plugin_result.verified_certificate_chain is None):
+            if cert_plugin_result.verified_certificate_chain is None:
                 logging.warning("{}: Untrusted certificate chain, probably due to missing intermediate certificate.".format(endpoint.url))
                 utils.debug("{}: Only {} certificates in certificate chain received.".format(endpoint.url, cert_plugin_result.received_certificate_chain.__len__()))
```
This needs to use `cert_plugin_result.certificate_deployments[]`. The `if` throws an error as it stands because the `CertificateInfoScanResult` class does not have a `verified_certificate_chain` attribute. The use of `cert_plugin_result.received_certificate_chain.__len__()` would throw a similar error.

mcdonnnj reviewed (changes requested):
Apologies for the delay in wrapping this up. Upon re-review and testing there are some problems that need to be fixed, and also some more questions for my team about multiple certificate deployments.

SaptakS committed:
Consider all certificate deployments for verified certificate chain

SaptakS commented:
@mcdonnnj @jsf9k I have tried to address the above reviews and comments to handle all certificate deployments. For verified_certificate_chain logics, I created a separate variable `has_verified_cert_chain` which is true only if all certificate deployments have a `verified_certificate_chain` and used that variable in the both the `if` logics. Also, I added the length of the `received_certificate_chain` for all the certificate deployments to get the `endpoint.https_cert_chain_len` and used the same in the error log since they were actually the same values it seems.

Let me know if all that sounds good.

SaptakS commented:
Hi @mcdonnnj did you get a chance to look at my latest changes?

eloquence commented:
Just a quick ping - would be great to get this over the finish line so we can update other dependencies like `cryptography` in our projects. (We may end up switching to a fork, at least temporarily, to do so.)

mcdonnnj commented:
@SaptakS @eloquence Sorry for the radio silence. I will re-review and test everything by Friday and hopefully we won't need any additional changes.

michaelblyons commented:
> I will re-review and test everything by Friday and hopefully we won't need any additional changes.

Is this a metaphysical "Friday" or an ordinary temporal "Friday?"

michaelblyons commented:
@mcdonnnj Nick, I beg you to spend some time on this.

mcdonnnj commented:
> @mcdonnnj Nick, I beg you to spend some time on this.

It's on the agenda but there have been other pressing development efforts for our team. Due to alignment with some internal needs it's hopefully becoming important enough that I can devote time to getting this situated and merged.

michaelblyons commented:
> Due to alignment with some internal needs it's hopefully becoming important enough that I can devote time to getting this situated and merged.

It looks like you talked them into PR #224 with a truly astounding 581 commits (25 new), doing `black` and such. But 5 months later, still no merge?

Even so, `sslyze` v2 [is still pinned][1] in that PR, when there's now a v5. We're 1.5 years from "test everything by Friday."

[1]: https://github.com/cisagov/pshtt/pull/224/files#diff-60f61ab7a8d1910d86d9fda2261620314edcae5894d5aaa236b821c7256badd7R100

epicfaace commented:
Note that this branch no longer works with sslyze 5. It only works with sslyze up to version 4.

With sslyze 5, I get this error:

```
$ pshtt example.com
Traceback (most recent call last):
  File "/home/codespace/.python/current/bin/pshtt", line 5, in <module>
    from pshtt.cli import main
  File "/usr/local/python/3.10.4/lib/python3.10/site-packages/pshtt/cli.py", line 27, in <module>
    from . import pshtt
  File "/usr/local/python/3.10.4/lib/python3.10/site-packages/pshtt/pshtt.py", line 30, in <module>
    from sslyze import (
ImportError: cannot import name 'ServerConnectivityTester' from 'sslyze' (/usr/local/python/3.10.4/lib/python3.10/site-packages/sslyze/__init__.py)
```

mcdonnnj commented:
I have merged #224, created the [`0.6.8` tag](https://github.com/cisagov/pshtt/releases/tag/v0.6.8), and published `0.6.8` on [PyPI](https://pypi.org/project/pshtt/0.6.8/). This pull request will need to be updated accordingly for us to proceed with a re-review, merge, and new release. There are some substantial changes made in the PR I mentioned so you may want to rebase or cherry pick.

Please note that the mentioned release will now allow you to install `pshtt` from PyPI on Python 3.7 without issue.

SaptakS committed:
Updates sslyze version to <5.0.0, upgrade required python to >=3.7

SaptakS committed:
Lints the python code

SaptakS commented:
@mcdonnnj I have rebased the PR on current develop and resolved all merge conflicts. Also, linted the code to match the best practices applied while skeletonizing of the project. I have changed the `python_requires` in setup.py because sslyze `>=3.x.x` doesn't support python 3.6 but supports everything above that. 

Also, pinned sslyze to `<5.0.0` because sslyze 5.0.0 changes the scanner API completely. So if we want to use sslyze 5.x.x, then we need to update the entire code and change sslyze requirement to `>=5.0.0`. I am not sure what you think is the best decision there.

SaptakS committed:
Updates python versions in github workflow build

mcdonnnj commented on setup.py:
```diff
@@ -77,15 +77,12 @@ def get_version(version_file):
         "Programming Language :: Python :: 3 :: Only",
         "Programming Language :: Python :: 3.6",
```
```suggestion
```

mcdonnnj reviewed (changes requested):
Just some quick housekeeping for the updated Python version support before I get into a proper review.

So that this is consistent with our other Python projects please change the `3.7` to `3.10` (and remove the comments) on the following lines:

https://github.com/freedomofpress/pshtt/blob/445131a4173fcf964ae1dee560c8281dc8228ab8/.github/workflows/build.yml#L26-L28
https://github.com/freedomofpress/pshtt/blob/445131a4173fcf964ae1dee560c8281dc8228ab8/.github/workflows/build.yml#L169-L171

SaptakS committed:
Updates python version in build.yml

SaptakS commented:
@mcdonnnj done!

epicfaace commented:
https://github.com/cisagov/scanner/pull/63#issuecomment-1335624713

michaelblyons commented:
![I am once again asking for someone to review and approve](https://user-images.githubusercontent.com/2701605/231167430-ac39e561-0453-466a-ae95-c38513086c63.png)

mcdonnnj commented:
@SaptakS Would you please rebase this on `develop` so we can look at getting this tested, approved, merged, and released?

jsf9k commented:
> @SaptakS Would you please rebase this on `develop` so we can look at getting this tested, approved, merged, and released?

I know that this PR has been egregiously neglected, but there is currently some interest from our leadership in this project.  We can finally devote some time to this; the iron is hot, so now is the time to strike.

SaptakS commented:
> I know that this PR has been egregiously neglected, but there is currently some interest from our leadership in this project. We can finally devote some time to this; the iron is hot, so now is the time to strike.

That sounds great! I am curious however that since [sslyze](https://github.com/nabla-c0d3/sslyze) has moved to 5.x.x. do we still want to update this PR such that it supports `sslyze>=3.0.0,<4` , or do we want to update this PR to use sslyze 5.x.x instead?

mcdonnnj commented:
> > I know that this PR has been egregiously neglected, but there is currently some interest from our leadership in this project. We can finally devote some time to this; the iron is hot, so now is the time to strike.
> 
> That sounds great! I am curious however that since [sslyze](https://github.com/nabla-c0d3/sslyze) has moved to 5.x.x. do we still want to update this PR such that it supports `sslyze>=3.0.0,<4` , or do we want to update this PR to use sslyze 5.x.x instead?

I would like to start with this PR as I'm relatively confident it will not have breaking behavior with our dependent projects (still need to verify that though). Supporting sslyze 5+ should be a separate effort.

SaptakS commented:
I have rebased, resolved conflicts and pushed. Waiting for the tests to pass. I think someone with permission need to approve the workflows.

mcdonnnj commented on .github/workflows/build.yml:
```diff
@@ -281,16 +255,9 @@ jobs:
+          - "3.9"
+          - "3.10"
           # - "3.11"
         include:
```
Please remove the matrix entry for Python 3.6 here since this PR changes support to Python 3.7+.

SaptakS replied:
Done.

SaptakS committed:
Removes Python 3.6 from build

michaelblyons commented:
Are the coveralls complaints innocuous?

mcdonnnj commented:
> Are the coveralls complaints innocuous?

Relatively. Coverage decreased because total lines of code increased, but the current configuration doesn't do extensive testing for what is being modified in the first place. It's unreasonable to ask that testing be added from scratch to cover these changes so manual testing will have to suffice. As a result the coveralls complaints will be ignored.

jsf9k reviewed (commented):
These changes seem correct to me, and I'm inclined to approve.  Have you done any testing beyond the (admittedly underwhelming) test suite?  What I mean is, have you been using these changes to test against a wide array of servers?

In any event, we should probably do some testing on our end, just to get a warm and fuzzy that everything is still working.  I can start looking into that next week.

harrislapiroff commented:
@jsf9k How's that looking? We're very interested in getting this merged :)

michaelblyons commented:
Hi folks. How was June? How's the review coming?

tgoetheyn commented:
Well, Python 3.7 has reached end-of-life.
so hopefully someone comes back from vacation with enough energy to finally tackle this pull request

gbinal commented:
Hi folks,  

Just to help nudge this, my team is trying to use pshtt right now and can't at all at this point (some of the second order dependencies appear to no longer be hosted anymore).  

Pretty please with sugar on top, can you all merge this PR?

jsf9k commented:
This needs another rebase on top of `develop` before it can be merged, but we (cisagov) don't have permission to push changes to the PR branch.  I did the rebase locally and it went cleanly, so this should be a five second task.

Alternatively, you can click the box to allow us to push up changes and I'll take care of it.

dav3r commented:
@SaptakS Can you also please bump the minor version by running this command?
```console
./bump_version.sh minor
```

This will need to be done prior to merging.

felddy commented:
![its-happening](https://github.com/cisagov/pshtt/assets/3229435/06777dfa-7ed2-4f2b-80a6-43dbd648f393)

felddy commented:
#248 was merged with a superset of these changes rebased to `develop`, as well as a few of the outstanding change requests.  I'm waiting on some "eventual consistency" with GitHub's view of this PR.  I think it will realize it is merged at some point.  If not, I'll come back and close it.

Thank you to everyone that contributed to the PR.
